### DISCUSSION TITLE
Segfault on glCreateShader in imgui_impl_opengl3.cpp

### DISCUSSION THREAD

**Discussion by IronFractal** [ocornut/imgui#5697] [New Users Build/Link/Run issues ONLY!]
imGui: 1.80 WIP 
OpenGL Version: 4.6 (Core Profile) Mesa 22.0.5
GLSL Version: 4.60
Rendered by: AMD Radeon RX 5500 XT (navi14, LLVM 13.0.1, DRM 3.47, 5.19.0-76051900-generic)
g_GlslVersionString:#version 460
backend: imgui_impl_opengl3.cpp
OS: Pop!_OS 22.04 LTS x86_64
Kernel: 5.19.0-76051900-generic

I am taking an intro to graphics course and was provided a template project with the following folder/file structure.
```
./
↳*.cpp
↳*.h
↳Makefile
↳lighting.vert
↳lighting.frag
↳... (other less important stuff like *.ply, textures folder with textures, etc...)
↳libs (dir)
  ↳glbinding (dir)
  ↳glw (dir)
  ↳glm (dir)
  ↳imgui-master (dir) [1.80 WIP repo]
  ↳...
```
*Note got a bunch of undefined references using the 1.88 release instead, but 1.80 no issues, so that is what I'm using.

Still not very familiar with all the libraries and what they do. I was able to compile the project with the provided Makefile.
```
g++ -MM -std=c++11 -g  -I. -Ilibs/glm -Ilibs/imgui-master -Ilibs/imgui-master/backends -Ilibs  -Ilibs/glfw/include -DVK_TAB=9 framework.cpp interact.cpp transform.cpp scene.cpp texture.cpp shapes.cpp object.cpp shader.cpp simplexnoise.cpp fbo.cpp emulator.cpp > dependencies
Compile framework.cpp
Compile interact.cpp
Compile transform.cpp
Compile scene.cpp
Compile texture.cpp
Compile shapes.cpp
Compile object.cpp
Compile shader.cpp
Compile simplexnoise.cpp
Compile fbo.cpp
Compile emulator.cpp
Compile libs/imgui-master/imgui.cpp
Compile libs/imgui-master/imgui_widgets.cpp
Compile libs/imgui-master/imgui_draw.cpp
Compile libs/imgui-master/imgui_demo.cpp
Compile libs/imgui-master/backends/imgui_impl_glfw.cpp
Compile libs/imgui-master/backends/imgui_impl_opengl3.cpp
Compile rply.c
Link bobjs/framework.exe
cd bobjs && g++ -g -o ../bobjs/framework.exe  framework.o interact.o transform.o scene.o texture.o shapes.o object.o shader.o simplexnoise.o fbo.o emulator.o imgui.o imgui_widgets.o imgui_draw.o imgui_demo.o imgui_impl_glfw.o imgui_impl_opengl3.o rply.o -std=c++11 -g  -I. -Ilibs/glm -Ilibs/imgui-master -Ilibs/imgui-master/backends -Ilibs  -Ilibs/glfw/include -DVK_TAB=9 -L/usr/lib/x86_64-linux-gnu -L../libs -L/usr/lib -L/usr/local/lib -lglbinding -lX11 -lGLEW -lGLU -lGL `pkg-config --static --libs glfw3`
```

The issue is when I try to run the resulting executable. I get a segmentation fault with the following stack trace.

```
main
↳Scene::DrawMenu
 ↳ImGui_ImplOpenGL3_NewFrame
  ↳ImGui_ImplOpenGL3_CreateDeviceObjects
   ↳glCreateShader (crash: SIGSEGV, Segmentation fault)
```

The issue appears to be that glCreateShader is null (address 0x0). My professor on the same OS is able to compile and run it with no issues (using Nvidia gpu instead [not sure if that has anything to do with it]). 

framework.h
```cpp
// All the includes neeed across several cpp files:
//   OpenGL, GLFW, GLM, IMGUI, etc

#include "imgui.h"
#include "imgui_impl_glfw.h"
#include "imgui_impl_opengl3.h"


#define GLFW_INCLUDE_NONE
#include <glbinding/Binding.h>  // Initialize with glbinding::Binding::initialize()
#include <glbinding/gl/gl.h>
using namespace gl;

#include <GLFW/glfw3.h>

#define GLM_FORCE_RADIANS
#define GLM_SWIZZLE
#include <glm/glm.hpp>

#include "shader.h"
#include "scene.h"
#include "interact.h"
```

framework.cpp (just main func [includes framework.h])
```cpp
int main(int argc, char** argv)
{
    glfwSetErrorCallback(error_callback);

    // Initialize the OpenGL bindings
    glbinding::Binding::initialize(false);

    // Initialize glfw open a window
    if (!glfwInit()) exit(EXIT_FAILURE);

    glfwWindowHint(GLFW_RESIZABLE, 1);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
    glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, 0);
    scene.window = glfwCreateWindow(750,750, "Graphics Framework", NULL, NULL);
    if (!scene.window)  { glfwTerminate();  exit(-1); }

    glfwMakeContextCurrent(scene.window);
    glfwSwapInterval(1);

    ImGui::CreateContext();
    //ImGuiIO& io = ImGui::GetIO(); (void)io;

    ImGui_ImplGlfw_InitForOpenGL(scene.window, true);
    ImGui_ImplOpenGL3_Init("#version 460");
    
    printf("OpenGL Version: %s\n", glGetString(GL_VERSION));
    printf("GLSL Version: %s\n", glGetString(GL_SHADING_LANGUAGE_VERSION));
    printf("Rendered by: %s\n", glGetString(GL_RENDERER));
    fflush(stdout);

    // Initialize interaction and the scene to be drawn.
    InitInteraction();
    scene.InitializeScene();

    // Enter the event loop.
    while (!glfwWindowShouldClose(scene.window)) {
        glfwPollEvents();

        scene.DrawScene();
        scene.DrawMenu();
        glfwSwapBuffers(scene.window); }

    ImGui_ImplOpenGL3_Shutdown();
    ImGui_ImplGlfw_Shutdown();
    ImGui::DestroyContext();

    glfwTerminate();
}
```

If I comment out the call to scene.DrawMenu(); it works just fine (just without imgui). DrawMenu is the function that draws the menu with imgui.

scene.cpp (just DrawMenu function)
```cpp
void Scene::DrawMenu()
{
    ImGui_ImplOpenGL3_NewFrame();
    ImGui_ImplGlfw_NewFrame();
    ImGui::NewFrame();

    if (ImGui::BeginMainMenuBar()) {
        // This menu demonstrates how to provide the user a list of toggleable settings.
        if (ImGui::BeginMenu("Objects")) {
            if (ImGui::MenuItem("Draw spheres", "", spheres->drawMe))  {spheres->drawMe ^= true; }
            if (ImGui::MenuItem("Draw walls", "", room->drawMe))       {room->drawMe ^= true; }
            if (ImGui::MenuItem("Draw ground/sea", "", ground->drawMe)){ground->drawMe ^= true;
                							sea->drawMe = ground->drawMe;}
            ImGui::EndMenu(); }
                	
        // This menu demonstrates how to provide the user a choice
        // among a set of choices.  The current choice is stored in a
        // variable named "mode" in the application, and sent to the
        // shader to be used as you wish.
        if (ImGui::BeginMenu("Menu ")) {
            if (ImGui::MenuItem("<sample menu of choices>", "",	false, false)) {}
            if (ImGui::MenuItem("Do nothing 0", "",		mode==0)) { mode=0; }
            if (ImGui::MenuItem("Do nothing 1", "",		mode==1)) { mode=1; }
            if (ImGui::MenuItem("Do nothing 2", "",		mode==2)) { mode=2; }
            ImGui::EndMenu(); }
        
        ImGui::EndMainMenuBar(); }
    glUseProgram(0);
    ImGui::Render();
    ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
}
```

Could not find any posts with similar issues.

**Comment by PathogenDavid** [comment]
Thanks for providing such detailed info and doing your due diligence. Unfortunately this sounds more like an OpenGL issue than a Dear ImGui one, but maybe I can get you headed in the right direction.

> The issue appears to be that glCreateShader is null (address 0x0). 

OpenGL functions being null usually indicates that you haven't initialized your OpenGL loader, or initializing it has failed. The fact you're getting it here is a little strange though since other OpenGL functions would've been called before `glCreateShader`.

> *Note got a bunch of undefined references using the 1.88 release instead, but 1.80 no issues, so that is what I'm using.

Ideally you should try to resolve this. In 1.84 we changed how Dear ImGui dealt with OpenGL loaders to help solve weird loader issues like this. (See https://github.com/ocornut/imgui/issues/4445 for details)

----------

You should also try to build the [official GLFW + OpenGL example](https://github.com/ocornut/imgui/tree/85f327d8d3710a32bbde08cb8995649ab257b765/examples/example_glfw_opengl3) in order to rule out issues specific to your project.

**Reply by IronFractal** [reply]
Thanks for your response. I forgot to mention in my original post, but I am able to compile that example without issue. 
![Screenshot from 2022-09-23 22-27-07](https://user-images.githubusercontent.com/20851279/192081543-31d1b471-99df-47ef-a251-a514dfad2389.png)
Also in my attempts to debug the issue I also placed some error checking code directly before the glCreateShader.
```C++
GLenum err = glGetError(); 
if (err != GL_NO_ERROR) 
{ fprintf(stderr, "OpenGL error (at line scene.cpp:%d): %s\n", __LINE__, gluErrorString(err)); 
}
```
With no error message printing (no OpenGL errors).
I would like to upgrade to the latest version, however I do not think that would be possible for me. I have to submit only the modified source files (.h, .cpp) back to my professor for grading, and any significant changes to the project / project dependencies would make it not compile. (needs to work with version 1.80)

**Reply by IronFractal** [reply]
I added the following line of code to directly before the while loop in my main function.
```C++
fprintf(stderr, "HERE: %d : %p\n", glCreateShader, glCreateShader);
```
And it prints the address of the glCreateShader function as `0x7f81069de370`, so it only appears to be null in the context of the backend imgui function.

**Reply by PathogenDavid** [reply]
Thanks for the additional info!

Looking at your build setup again, one thing that stands out to me is that you have two different OpenGL loaders. (`GLEW` and `glbinding`) I wonder if your code is using one and Dear ImGui is using the other, and only the other is failing to initialize properly. (Having `glCreateShader` be null in the backend and non-null in your code would definitely point towards that.)

I don't see any of the `IMGUI_IMPL_OPENGL_LOADER_*` being defined. Do you have any customizations in your `imconfig.h`? (`imconfig.h` should be all comments by default, so anything that isn't is a customization.)

**Reply by IronFractal** [reply]
Yes, the `imconfig.h` file appears to only consist entirely of comments. Unchanged from the default.

**Comment by ocornut** [comment]
You haven’t initialized the loader used by the imgui_impl_opengl3.cpp file (gl3w by default), but you can probably only do it from a compilation unit that doesn’t already include a loader.

A possible workaround is to #include that cpp file from an existing cpp file of yours that has your loader included. Tho really you should mostly just update the backends and/or the whole of dear imgui.

**Reply by IronFractal** [reply]
I first attempted your workaround solution, but could not get it to work. So I ended up trying to compile with 1.88 again and realized that the issue was that the Makefile did not list `imgui_tables.cpp` and thus it was not compiled / linked. Was `imgui_tables.cpp` a new addition / refactor since 1.80? After adding that cpp to the list of Imgui sources, it compiles and runs just fine with 1.88. 
![Screenshot from 2022-09-27 22-46-39](https://user-images.githubusercontent.com/20851279/192697636-52b6c5b5-cd3b-4837-aa4e-2793bb5187e0.png)
The menu does not respond to interaction (mouse clicks etc…) but I think that may be due to something on my project's end not being implemented? Our first assignment is to implement interaction with mouse buttons, so I'm fairly sure that is the reason.

**Reply by IronFractal** [reply]
I figured out why the menu buttons were not responding to interaction. The code I am working with calls some callback setup functions that seem to overide the imgui backend glfw code. 

```c++
glfwSetKeyCallback(scene.window, Keyboard);
glfwSetMouseButtonCallback(scene.window, MouseButton);
glfwSetCursorPosCallback(scene.window, MouseMotion);
glfwSetScrollCallback(scene.window, Scroll);
```

These callback functions check `ImGui::GetIO().WantCaptureMouse` and simply return when true. 
```c++
if (ImGui::GetIO().WantCaptureMouse) return;
```

I found the following post https://github.com/ocornut/imgui/issues/1759. I added calls to the ImGui callback function if `WantCaptureMouse` is true.
```c++
if (ImGui::GetIO().WantCaptureMouse)
{

    ImGui_ImplGlfw_ScrollCallback(window, x, y);

    return;

}
```
```c++
if (ImGui::GetIO().WantCaptureMouse)
{

    ImGui_ImplGlfw_CursorPosCallback(window, x, y);

    return;

}
```
```c++
if (ImGui::GetIO().WantCaptureMouse)
{

    ImGui_ImplGlfw_MouseButtonCallback(window, button, action, mods);

    return;

}
```
```c++
if (ImGui::GetIO().WantCaptureKeyboard)
{

    ImGui_ImplGlfw_KeyCallback(window, key, scancode, action, mods);

    return;

}
```

This fixed a lot of the bugs I was having with the menu not detecting mouse hover / mouse click on menu buttons / items. However, I am still getting issues with this callback.
```c++
/* \/ \/ \/ Problematic */
glfwSetCursorPosCallback(scene.window, MouseMotion);
```
With this one commented out, everything works fine as expected. With it not commented out, the menu does not detect hover / mouse clicked unless I move the window. Then the menu becomes useable. It will stay useable until I change window's size. Do you have an idea as to what may be causing this?

**Comment by ocornut** [comment]
If you install your callbacks BEFORE installing the imgui_impl_glfw callbacks, then they will automatically be chained and you don’t need to do the chaining.
